src/lib.rs

```diff
@@ -13,24 +13,103 @@
+pub const TTL: Timespec = Timespec { sec: 60, nsec: 0 };
 
 /// FUSE file system implementation of sandboxfs.
 struct SandboxFS {
+    /// Mapping of inode numbers to in-memory nodes that tracks all files known by sandboxfs.
+    nodes: Arc<Mutex<HashMap<u64, Arc<nodes::Node>>>>,
 }
 
 impl SandboxFS {
     /// Creates a new `SandboxFS` instance.
     fn new() -> SandboxFS {
-        SandboxFS {}
+        let root = {
+            let inode = fuse::FUSE_ROOT_ID;
+            let now = time::get_time();
+            let uid = unsafe { libc::getuid() } as u32;
+            let gid = unsafe { libc::getgid() } as u32;
+            nodes::Dir::new_empty(inode, inode, now, uid, gid)
+        };
+
+        let mut nodes = HashMap::new();
+        nodes.insert(root.inode(), root);
+        SandboxFS {
+            nodes: Arc::from(Mutex::from(nodes)),
+        }
+    }
+
+    /// Gets a node given its `inode`.
+    ///
+    /// We assume that the inode number is valid and that we have a known node for it; otherwise,
+    /// we crash.  The rationale for this is that this function is always called on inode numbers
+    /// requested by the kernel, and we can trust that the kernel will only ever ask us for inode
+    /// numbers we have previously told it about.
+    fn find_node(&mut self, inode: u64) -> Arc<nodes::Node> {
+        let nodes = self.nodes.lock().unwrap();
+        match nodes.get(&inode) {
+            Some(node) => node.clone(),
+            None => panic!("Kernel requested unknown inode {}", inode),
+        }
     }
 }
 
 impl fuse::Filesystem for SandboxFS {
+    fn getattr(&mut self, _req: &fuse::Request, inode: u64, reply: fuse::ReplyAttr) {
+        let node = self.find_node(inode);
+        match node.getattr() {
+            Ok(attr) => reply.attr(&TTL, &attr),
+            Err(e) => reply.error(e.raw_os_error().unwrap()),
+        }
+    }
+
+    fn lookup(&mut self, _req: &fuse::Request, parent: u64, name: &OsStr, reply: fuse::ReplyEntry) {
+        let dir_node = self.find_node(parent);
+        match dir_node.lookup(name) {
+            Ok((node, attr)) => {
+                {
+                    let mut nodes = self.nodes.lock().unwrap();
+                    if !nodes.contains_key(&node.inode()) {
+                        nodes.insert(node.inode(), node);
+                    }
+                }
+                reply.entry(&TTL, &attr, 0);
+            },
+            Err(e) => reply.error(e.raw_os_error().unwrap()),
+        }
+    }
+
+    fn readdir(&mut self, _req: &fuse::Request, inode: u64, _handle: u64, offset: i64,
+               mut reply: fuse::ReplyDirectory) {
+        if offset == 0 {
+            let node = self.find_node(inode);
+            match node.readdir(&mut reply) {
+                Ok(()) => reply.ok(),
+                Err(e) => reply.error(e.raw_os_error().unwrap()),
+            }
+        } else {
+            assert!(offset > 0, "Do not know what to do with a negative offset");
```

Comment:
More reasonable to return an error than to die here. If we had logging I'd suggest logging an error as well.

Reply:
The value we get here as an offset is a kernel construct–the `readdir` system call that the user invokes has no concept of offset). We have to trust the kernel to send us valid data here, which means a positive or zero offset.

If we get a negative value, it either implies there is a bug in the kernel or, more likely, that we haven't thought about some case that has to be handled. Either way, we don't know why that is, which means our mental model of this function doesn't match reality. I'd rather crash in those situations than to proceed with a potential inconsistency...